### ISSUE TITLE
Initial sync may deadlock

### ISSUE THREAD

**Issue by anatomybook** [aws-amplify/amplify-swift#1865]
### Describe the bug

When running the initial sync, the app can deadlock. Interestingly enough, I'm currently only seeing this, when putting breakpoints in certain positions. I believe, however, that if it can be triggered with the debugger, it can also happen without.

### Steps To Reproduce

```swift
Steps to reproduce the behavior:
1. Erase App
2. Place breakpoints in `ModelSyncedEventEmitter`
2.1. Line 119, `initialSyncOperationFinished = true`
2.2. Line 156, `sendModelSyncedEvent()`
2.3. Line 162, `sendModelSyncedEvent()`
3. Start App, triggering sync from scratch
4. As soon as debugger stops, click continue
5. App freezes completely.
```

### Expected behavior

I expect the App to never freeze.

### Amplify Framework Version

1.26.1

### Amplify Categories

DataStore

### Dependency manager

Cocoapods

### Swift version

swift-driver version: 1.45.2 Apple Swift version 5.6 (swiftlang-5.6.0.323.62 clang-1316.0.20.8)

### CLI version

8.2.0

### Xcode version

Xcode 13.3.1, Build version 13E500a

### Is this a regression?

No

### Device

Simulator

### iOS Version

15.4

### Specific to simulators

unknown

### Additional context

Here is where the threads are stuck:
Main:
![Screenshot 2022-06-05 at 09 33 52](https://user-images.githubusercontent.com/67764399/172040476-65f4f3c2-8366-4796-a517-e84473191cd1.png)

Amazon Threads:
![Screenshot 2022-06-05 at 09 35 52](https://user-images.githubusercontent.com/67764399/172040511-42054576-12ac-4f64-8c6f-8fb2cfc9746f.png)

![Screenshot 2022-06-05 at 09 39 34](https://user-images.githubusercontent.com/67764399/172040640-433a75d4-f7df-4b2c-985d-fb3e76b62f33.png)

**Comment by atierian** [comment]
Thanks for opening this @anatomybook. 
Could you please post the stack trace in text form? You can copy and paste it from the Debug Navigator where you took those screenshots. Thanks!

**Comment by atierian** [comment]
Thanks for this thorough report @anatomybook, it's very helpful. We'll update here with any news.

**Comment by lawmicha** [comment]
Hi @anatomybook, really appreciate the level of details you've provided and your patience. Regarding the change in https://github.com/aws-amplify/amplify-ios/issues/1864 to make sure ModelSyncedEvent is emitted, it look like the correct change. I cut the PR for that issue in https://github.com/aws-amplify/amplify-ios/pull/1989 under the branch `fix/modelsynced-sync-finished`. Since that PR fix has your code change, we can continue to debug the deadlock issue you're seeing with this PR branch (while it's in PR), to avoid confusion. 

The PR adds your code change, then as a first step towards investigating the deadlock issue, i've moved the queue off of the global queue in the second commit (https://github.com/aws-amplify/amplify-ios/pull/1989/commits/96804bca6aae377ff8f59be00322437a57a2d74a). I'm not 100% if this is the problem but I recall discussing this topic with @atierian briefly about stop using the global queue so we wanted to clean up the code around this anyways. 

Glancing over the code, the intention of the private queue in ModelSyncedEventEmitter is to serialize the asynchronous work of processing both the sync operation events and reconciliation events. There are even variables like `dispatchedModelSyncedEvent` that is just a bool that is made thread safe because of the access to it is made only under the work put on the queue. 

Do you have a schema (redacted if needed) and expected number of models persisted that you can provide us to replicate a similar environment to reproduce the issue?

**Comment by anatomybook** [comment]
Hello @lawmicha ,

sorry for my delayed reply. Here's my schema - slightly edited:
```
type ActorZ @model @auth(rules: [{ allow: owner }])
{
  id: ID!
  .
  .
  .
  myContents: [ContentsZ] @connection(name: "LinkToActor")
  groups: [GroupZ] @connection(name: "LinkToGroups")
  myLists: [ListZ] @connection(name: "LinkToLists")
}

type ContentsZ @model @auth(rules: [{ allow: owner }])
{
  id: ID!
  creation: AWSDateTime!
  dateOnReceipt: AWSDateTime!
  .
  .
  .
  actor: ActorZ @connection(name: "LinkToActor")
  contents: [ContentsFieldZ] @connection(name: "LinkToContent")
  filenames: [ImageZ] @connection(name: "LinkToImages")
}

type ImageZ @model @auth(rules: [{ allow: owner }])
{
  id: ID!
  sThreeKey: String!
  lastLookedAt: AWSDateTime
  remarks: String
  details: String
  receipt: ContentsZ @connection(name: "LinkToImages")
}

type ContentsFieldZ @model @auth(rules: [{ allow: owner }])
{
  id: ID!
  caption: String!
  orderByMe: Int!
  price: Int
  .
  .
  .
  .
  .
  .
  .
  .
  .
  .
  parent: ContentsZ @connection(name: "LinkToContent")
  groups: [ContentsFieldVsGroupZ] @connection(keyName: "byContentsField", fields: ["id"])
}

type ContentsFieldVsGroupZ @model(queries: null) @auth(rules: [{ allow: owner }])
@key(name: "byContentsField", fields: ["myContentsFieldId", "groupId"])
@key(name: "byGroupA", fields: ["groupId", "myContentsFieldId"])
{
  id: ID!
  myContentsFieldId: ID!
  groupId: ID!
  myContentsField: ContentsFieldZ! @connection(fields: ["myContentsFieldId"])
  group: GroupZ! @connection(fields: ["groupId"])
}

type GroupZ @model @auth(rules: [{ allow: owner }])
{
  id: ID!
  .
  .
  .
  .
  .
  .
  .
  .
  .
  .
  actor: ActorZ @connection(name: "LinkToGroups")
  searches: [SearchZ] @connection(name: "LinkToSearches")
  prognosis: [PrognosisZ] @connection(name: "LinkToPrognosis")
  statusChangeTime: [StatusChangeTimeZ] @connection(name: "LinkToInCartTime")
  myLists: [ListVsGroupZ] @connection(keyName: "byGroup", fields: ["id"])
  myContentsFields: [ContentsFieldVsGroupZ] @connection(keyName: "byGroupA", fields: ["id"])
}

type SearchZ @model @auth(rules: [{ allow: owner }])
{
  id: ID!
  phrase: String
  comment: String
  .
  .
  .
  .
  .
  .
  threshold: Float
  group: GroupZ @connection(name: "LinkToSearches")
}

type StatusChangeTimeZ @model @auth(rules: [{ allow: owner }])
{
  id: ID!
  .
  statusChangeTime: AWSDateTime!
  group: GroupZ @connection(name: "LinkToInCartTime")
}

type PrognosisZ @model @auth(rules: [{ allow: owner }])
{
  id: ID!
  expected: Float!
  lower: Float!
  upper: Float!
  updated: AWSDateTime!
  .
  group: GroupZ @connection(name: "LinkToPrognosis")
}

type ListVsGroupZ @model(queries: null) @auth(rules: [{ allow: owner }])
@key(name: "byList", fields: ["myListId", "groupId"])
@key(name: "byGroup", fields: ["groupId", "myListId"])
{
  id: ID!
  amount: Float
  .
  .
  .
  .
  .
  .
  .
  myListId: ID!
  groupId: ID!
  myList: ListZ! @connection(fields: ["myListId"])
  group: GroupZ! @connection(fields: ["groupId"])
}

type ListZ @model @auth(rules: [{ allow: owner }])
{
  id: ID!
  name: String!
  thresholdPercentage: Float
  imageData: String
  imageDetails: String
  usageDetails: String
  .
  .
  .
  .
  creation: AWSDateTime!
  actor: ActorZ @connection(name: "LinkToLists")
  groups: [ListVsGroupZ] @connection(keyName: "byList", fields: ["id"])
}

```

The number of items is quite low in this user-account:
1 x ActorZ
6 x StatusChangeTimeZ
5 x GroupZ
8 x ContentsFieldVsGroupZ
102 x ContentsFieldZ
2 x ContentsZ
2 x ImageZ
5 x SearchZ
4 x ListVsGroupZ
1 x ListZ
0 x PrognosisZ

I hope this helps. I'll check out the deadlock tomorrow.

Thanks for your support!

**Comment by anatomybook** [comment]
> Could you try and reproduce the deadlock again using `fix/modelsynced-sync-finished` branch?

I just tried. Everything works - no more deadlock! It'd be awesome if that made it into main. 

Thank you every much.

**Comment by lawmicha** [comment]
Hi @anatomybook thanks for your patience, `fix/modelsynced-sync-finished` changes were merged in another PR that made it to main and just released in 1.28.0 https://github.com/aws-amplify/amplify-ios/releases/tag/v1.28.0 

Keeping this issue open to look into the schema you've provided

**Comment by anatomybook** [comment]
> Keeping this issue open to look into the schema you've provided

@lawmicha - thanks for the inclusion of my suggestion.

Please let me know if you have any questions or need more details regarding the schema. Talking about - the schema is old. I haven't touched it in ages. Given that I started early during the development of DataStore, I'm sometimes wondering whether the issues I'm seeing are related to it being old and never updated. In case you have any hints - perhaps new and special keywords - that I may incorporate, please let me know. I haven't found anything, but perhaps I'm overlooking things.

Two headaches this schema still gives me: I'm currently still struggling with performance, which I've tracked down to this issue: https://github.com/aws-amplify/amplify-ios/issues/1579. Additionally, I'm seeing orphaned children quite often. https://github.com/aws-amplify/amplify-ios/issues/2059.
